### PR TITLE
add timestamp constant function;

### LINKED ISSUE #2759: when there is no constant function,the predicate push down failed
when there is no constant function,such as the timestamp,str_to_date(date_format(`t0`.`__fcol_13`, '%Y-%m-%d'), '%Y-%m-%d'),date,the predicate push down failed,the starrocks scan all data

### PR DESCRIPTION
**Pull request by bigdata-kuxingseng**
Fix #2759
This PR contains 3 parts of optimization.
1.add timestamp constant function
2.modify millisecond analysis function
3.optimize datetime cast situation

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java [hunk 1]
```diff
@@ -49,9 +53,22 @@
...
+
     private static final LocalDateTime MAX_DATETIME = LocalDateTime.of(9999, 12, 31, 23, 59, 59);
     private static final LocalDateTime MIN_DATETIME = LocalDateTime.of(0, 1, 1, 0, 0, 0);
 
+    private static java.time.format.DateTimeFormatter DATE_TIME_FORMATTER_MS = null;
+
+    static {
+        try {
+            DATE_TIME_FORMATTER_MS = ScalarOperatorFunctions.unixDatetimeFormatBuilder("%Y-%m-%d %H:%i:%s.%f").toFormatter();
+        } catch (Exception e) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java:66]: This catch can be removed

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java [hunk 2]
```diff
@@ -67,6 +69,17 @@ public ScalarOperator visitBinaryPredicate(BinaryPredicateOperator operator,
         }
 
         ScalarOperator castChild = child1.getChild(0);
+        boolean isConstantOperator = child2 instanceof ConstantOperator;
+
+        // abandon cast function when cast datetime to date
+        if (child1 instanceof CastOperator && isConstantOperator) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java:75]: already checked that before

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java [hunk 3]
```diff
@@ -104,4 +117,66 @@ public boolean checkCastTypeReduceAble(Type parent, Type child, Type grandChild)
...
+
+        ScalarOperator castChild = child1.getChild(0);
+        // optimize cast
+        LocalDateTime originalDateTime = ((ConstantOperator) child2).getDatetime();
+        LocalDateTime bottomLocalDateTime = ((ConstantOperator) child2).getDatetime().toLocalDate().atTime(0, 0, 0, 0);
+        LocalDateTime targetLocalDateTime;
+        BinaryPredicateOperator.BinaryType binaryType = operator.getBinaryType();
+        int offset;
+        BinaryPredicateOperator.BinaryType newBinaryType;
+        if (binaryType.equals(BinaryPredicateOperator.BinaryType.GE)) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java:136]: plz add some comment like example (from old binary predicate to new binary predicate )to help people understand

**Comment by Seaven** [comment]
Rebase code plz

**Comment by Astralidea** [comment]
run all

**Comment by Astralidea** [comment]
run starrocks_be_unittest

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java [hunk 4]
```diff
@@ -444,4 +451,110 @@ private static ConstantOperator createDecimalConstant(BigDecimal result) {
         return ConstantOperator.createDecimal(result, type);
     }
 
+    public static DateTimeFormatterBuilder unixDatetimeFormatBuilder(String pattern) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java:454]: The function was move to DateUtils.java, don't need add again

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java [hunk 5]
```diff
@@ -115,7 +117,7 @@ public static ConstantOperator dateFormat(ConstantOperator date, ConstantOperato
         String format = fmtLiteral.getVarchar();
         // unix style
         if (!SUPPORT_JAVA_STYLE_DATETIME_FORMATTER.contains(format.trim())) {
-            DateTimeFormatterBuilder builder = DateUtils.unixDatetimeFormatBuilder(fmtLiteral.getVarchar());
+            DateTimeFormatterBuilder builder = unixDatetimeFormatBuilder(fmtLiteral.getVarchar());
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java:120]: Don't modify it

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java [hunk 6]
```diff
@@ -49,9 +51,13 @@
  * TYPE_PERCENTILE |    NOT_SUPPORT
  */
 public final class ConstantOperator extends ScalarOperator implements Comparable<ConstantOperator> {
+
     private static final LocalDateTime MAX_DATETIME = LocalDateTime.of(9999, 12, 31, 23, 59, 59);
     private static final LocalDateTime MIN_DATETIME = LocalDateTime.of(0, 1, 1, 0, 0, 0);
 
+    private static java.time.format.DateTimeFormatter DATE_TIME_FORMATTER_MS
+            = ScalarOperatorFunctions.unixDatetimeFormatBuilder("%Y-%m-%d %H:%i:%s.%f").toFormatter();
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java:59]:
```suggestion
    private static DateTimeFormatter DATE_TIME_FORMATTER_MS
            = DateUtils.unixDatetimeFormatBuilder("%Y-%m-%d %H:%i:%s.%f").toFormatter();
```

**Review by Seaven — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java [hunk 7]
```diff
@@ -104,4 +117,79 @@ public boolean checkCastTypeReduceAble(Type parent, Type child, Type grandChild)
...
+            offset = 0;
+            targetLocalDateTime = bottomLocalDateTime.plusDays(offset);
+            ConstantOperator newDate = ConstantOperator.createDate(targetLocalDateTime.truncatedTo(ChronoUnit.DAYS));
+            BinaryPredicateOperator binaryPredicateOperator =
+                    new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.LE, castChild, newDate);
+            return binaryPredicateOperator;
+        } else if (binaryType.equals(BinaryPredicateOperator.BinaryType.LT)) {
+            // when the BinaryType is < ,cast dateTime to maximum date type；
+            // Eg:cast dateTime(2021-12-28 00:00:00.0) to date(2021-12-27)
+            // Eg:cast dateTime(2021-12-28 00:00:00.1) to date(2021-12-28)
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java:172]: This offset is -1,  dateTime(2021-12-28 00:00:00.1) will cast to date(2021-12-27).

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java [hunk 8]
```diff
@@ -67,6 +69,17 @@ public ScalarOperator visitBinaryPredicate(BinaryPredicateOperator operator,
         }
 
         ScalarOperator castChild = child1.getChild(0);
+        boolean isConstantOperator = child2 instanceof ConstantOperator;
+
+        // abandon cast function when cast datetime to date
+        if (child1 instanceof CastOperator && isConstantOperator) {
+            boolean isCastFunction = castChild instanceof CastOperator;
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java:76]: Why need check the castOperator child is CastOperator?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java [hunk 9]
```diff
@@ -104,4 +117,79 @@ public boolean checkCastTypeReduceAble(Type parent, Type child, Type grandChild)
         Type parentCompatibleType = Type.getAssignmentCompatibleType(child, parent, true);
         return childCompatibleType != Type.INVALID && parentCompatibleType != Type.INVALID;
     }
+
+    public ScalarOperator optimizeDateTimeToDateCastSituation(BinaryPredicateOperator operator) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java:121]: How about this 
```
public ScalarOperator optimizeDateTimeToDateCastSituation(ConstantOperator leftChild,  ConstantOperator rightChild)
```
There will no need to chceck the child1 and child2 again

**Comment by dirtysalt** [comment]
@bigdata-kuxingseng  could you be kind to consider this PR? https://github.com/StarRocks/starrocks/pull/3343/files

I think this PR can help your code path

```
        } else {
            // current not support !=
            return operator;
        }
```

For some special literals like '2021-01-01 00:00:00', maybe you don't need to consider binary operator at all.

BTW, great work. Your PR is more thorough & well-thought than mine.  Thanks.

**Review by Youngwb — DISMISSED** [review]

**Review by Seaven — DISMISSED** [review]

**Review by Youngwb — DISMISSED** [review]

**Review by Seaven — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java [hunk 10]
```diff
@@ -104,4 +112,82 @@ public boolean checkCastTypeReduceAble(Type parent, Type child, Type grandChild)
...
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java:116]:
```suggestion
    public ScalarOperator optimizeDateTimeToDateCast(BinaryPredicateOperator operator) {
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java [hunk 11]
```diff
@@ -104,4 +112,82 @@ public boolean checkCastTypeReduceAble(Type parent, Type child, Type grandChild)
...
+        ScalarOperator child1 = operator.getChild(0);
+        ScalarOperator child2 = operator.getChild(1);
+
+        ScalarOperator castChild = child1.getChild(0);
+        // optimize cast
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java:121]: remove unused comment

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java [hunk 12]
```diff
@@ -104,4 +112,82 @@ public boolean checkCastTypeReduceAble(Type parent, Type child, Type grandChild)
...
+        LocalDateTime originalDateTime = ((ConstantOperator) child2).getDatetime();
+        LocalDateTime bottomLocalDateTime = ((ConstantOperator) child2).getDatetime().toLocalDate().atTime(0, 0, 0, 0);
+        LocalDateTime targetLocalDateTime;
+        BinaryPredicateOperator.BinaryType binaryType = operator.getBinaryType();
+        int offset;
+        if (binaryType.equals(BinaryPredicateOperator.BinaryType.GE)) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java:127]: Would better use switch case.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java [hunk 13]
```diff
@@ -104,4 +112,82 @@ public boolean checkCastTypeReduceAble(Type parent, Type child, Type grandChild)
...
+            targetLocalDateTime = bottomLocalDateTime.plusDays(offset);
+            ConstantOperator newDate = ConstantOperator.createDate(targetLocalDateTime.truncatedTo(ChronoUnit.DAYS));
+            BinaryPredicateOperator binaryPredicateOperator =
+                    new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.GE, castChild, newDate);
+            return binaryPredicateOperator;
+        } else if (binaryType.equals(BinaryPredicateOperator.BinaryType.LE)) {
+            // when the BinaryType is <= ,cast dateTime to maximum date type；
+            // Eg:cast dateTime(2021-12-28 00:00:00.0) to date(2021-12-28)
+            // Eg:cast dateTime(2021-12-28 00:00:00.1) to date(2021-12-28)
+            // Eg:cast dateTime(2021-12-28 00:20:00.0) to date(2021-12-28)
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java:155]: redundant comment

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java [hunk 14]
```diff
@@ -104,4 +112,82 @@ public boolean checkCastTypeReduceAble(Type parent, Type child, Type grandChild)
...
+            targetLocalDateTime = bottomLocalDateTime.plusDays(offset);
+            ConstantOperator newDate = ConstantOperator.createDate(targetLocalDateTime.truncatedTo(ChronoUnit.DAYS));
+            BinaryPredicateOperator binaryPredicateOperator =
+                    new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.LE, castChild, newDate);
+            return binaryPredicateOperator;
+        } else if (binaryType.equals(BinaryPredicateOperator.BinaryType.LT)) {
+            // when the BinaryType is < ,cast dateTime to maximum date type；
+            // Eg:cast dateTime(2021-12-28 00:00:00.0) to date(2021-12-27)
+            // Eg:cast dateTime(2021-12-28 00:00:00.1) to date(2021-12-28)
+            // Eg:cast dateTime(2021-12-28 12:00:00.1) to date(2021-12-28)
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java:166]: redundant comment

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/DistributedEnvPlanWithCostTest.java [hunk 15]
```diff
@@ -1120,6 +1120,37 @@ public void testCastPartitionPrune() throws Exception {
...
+        Assert.assertTrue(plan.contains("partitions=2/3"));
+    }
+
+    @Test
+    public void testTimestampPartitionPruneOptimize() throws Exception {
+        String sql =
+                "select * from test_partition_prune_optimize_by_date where dt >= timestamp('2021-12-27 00:00:00.123456') " +
+                        "and dt <= timestamp('2021-12-29 00:00:00.123456')";
+        String plan = getFragmentPlan(sql);
+        // check not prune partition
```
Comment (Youngwb) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/DistributedEnvPlanWithCostTest.java:1132]: give right annotation

**Review by kangkaisen — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Pull request merged**